runtime/jsTest/src/kotlinx/serialization/DynamicSerializerTest.kt

```diff
@@ -0,0 +1,205 @@
+package kotlinx.serialization
+
+import kotlinx.serialization.internal.IntArraySerializer
+import kotlinx.serialization.internal.IntSerializer
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.JsonConfiguration
+import kotlinx.serialization.modules.EmptyModule
+import kotlinx.serialization.modules.getContextualOrDefault
+import kotlinx.serialization.test.CommonEnumSerializer
+import kotlin.test.*
+
+
+class DynamicSerializerTest {
+    @Serializable
+    data class Data(val a: Int)
+
+    @Serializable
+    open class DataWrapper(open val s: String, val d: Data? = Data(1))
+
+    @Serializable
+    data class NestedList(val a: String, val list: List<Int>)
+
+    @Serializable
+    data class ListOfLists(val l: List<List<Data>>)
+
+    @Serializable
+    data class MapWrapper(val m: Map<String, Int>)
+
+    @Serializable
+    data class ComplexMapWrapper(val m: Map<String, Data>)
+
+    @Serializable
+    data class WithChar(val a: Char)
+
+    @Serializable
+    data class AllTypes(
+        val b: Byte,
+        val s: Short,
+        val i: Int,
+        val f: Float,
+        val d: Double,
+        val c: Char,
+        val B: Boolean,
+        val S: String
+    )
+
+    @Serializable
+    data class EnumWrapper(val e: Color)
+
+    enum class Color {
+        RED, GREEN
```

Comment:
Can we add some `@SerialId` and `@SerialName` test cases to make sure we cover those situations, especially with enums where it's easy to accidentally use `name`?